### PR TITLE
Include per-class AP75 in validation summaries and exports

### PR DESCRIPTION
**Pull request by woreom**
Per-class validation summaries omit AP75 even though it is already computed. Add `mAP75` from `self.box.all_ap[i, 5]` to `DetMetrics.summary()`, which also exposes it through the existing DataFrame, CSV, and JSON exports and inherited box summaries for segmentation, pose, and OBB.

The change is one added line. It adds an export column without changing metric tuple lengths, training-log keys, or introducing configuration arguments.

Validation: real YOLO26n detection, segmentation, pose, and OBB validation on COCO8/COCO8-seg/COCO8-pose/DOTA8 at imgsz=320; every per-class AP75 value matched the underlying IoU=0.75 column in summary, DataFrame, CSV, and JSON. Empty summaries passed for all four tasks. Ruff check and format check passed.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @woreom, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Please review the checklist below for a smooth integration 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/metrics.py [hunk 1]
```diff
@@ -1140,18 +1143,33 @@ def clear_image_metrics(self) -> None:
...
         """Return a list of keys for accessing specific metrics."""
-        return ["metrics/precision(B)", "metrics/recall(B)", "metrics/mAP50(B)", "metrics/mAP50-95(B)"]
+        base = ["metrics/precision(B)", "metrics/recall(B)", "metrics/mAP50(B)", "metrics/mAP50-95(B)"]
+        return base + [f"metrics/mAP{round(t * 100)}(B)" for t in self.iou_metrics]
 
     def mean_results(self) -> list[float]:
-        """Calculate mean of detected objects & return precision, recall, mAP50, and mAP50-95."""
-        return self.box.mean_results()
+        """Calculate mean of detected objects & return precision, recall, mAP50, mAP50-95, and any extra IoU metrics."""
+        return self.box.mean_results() + self._extra_maps()
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/metrics.py]: ⚠️ **HIGH**: `mean_results()` now returns more than 4 metric values when `iou_metrics` is set, but this PR does not update the row formatter in `DetectionValidator.print_results()` to add matching numeric placeholders. With extra thresholds enabled, the existing `%` formatting will receive too many arguments and fail while logging the summary/per-class rows.

File: ultralytics/models/yolo/detect/val.py [hunk 2]
```diff
@@ -289,6 +291,12 @@ def print_results(self) -> None:
                     )
                 )
 
+        # Save per-class metrics CSV when requested
+        if not self.training and getattr(self.args, "save_csv", False) and self.nc > 1 and len(self.metrics.stats):
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/val.py]: 💡 **MEDIUM**: `len(self.metrics.stats)` is always truthy here because `stats` is a dict with fixed keys, so this guard does not actually check whether validation produced any per-class data. On empty/degenerate runs this will still try to export a CSV, which is likely not what was intended.

**Suggested change:**
```suggestion
        if not self.training and getattr(self.args, "save_csv", False) and self.nc > 1 and any(len(v) for v in self.metrics.stats.values()):
```

File: ultralytics/utils/metrics.py [hunk 3]
```diff
@@ -1140,18 +1143,33 @@ def clear_image_metrics(self) -> None:
...
+        if not self.iou_metrics or not len(self.box.all_ap):
+            return []
+        return [float(self.box.all_ap[:, self._iou_idx(t)].mean()) for t in self.iou_metrics]
+
     @property
     def keys(self) -> list[str]:
         """Return a list of keys for accessing specific metrics."""
-        return ["metrics/precision(B)", "metrics/recall(B)", "metrics/mAP50(B)", "metrics/mAP50-95(B)"]
+        base = ["metrics/precision(B)", "metrics/recall(B)", "metrics/mAP50(B)", "metrics/mAP50-95(B)"]
+        return base + [f"metrics/mAP{round(t * 100)}(B)" for t in self.iou_metrics]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/metrics.py]: 💡 **MEDIUM**: The exported key names are built from the raw user input, but `_iou_idx()` silently rounds/clamps to the nearest supported evaluation bin. For example, `0.76` would be reported as `metrics/mAP76(B)` even though the value actually comes from the `0.75` column. That makes the output misleading unless unsupported thresholds are rejected or the labels are normalized to the snapped grid value.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The feature wiring looks straightforward, but there are a few correctness issues to address before merging: enabling `iou_metrics` appears to break console formatting, arbitrary thresholds can be mislabeled in outputs, and the CSV export guard currently always passes. Fixing those should make the PR solid.

💬 Posted 3 inline comments

File: tests/test_python.py [hunk 4]
```diff
@@ -235,6 +235,30 @@ def test_val(task: str, weight: str, data: str) -> None:
...
+    assert "metrics/mAP95(B)" in metrics.results_dict, "mAP95 key missing from results_dict"
+    # Standard keys must still be present
+    assert "metrics/mAP50(B)" in metrics.results_dict
+    assert "metrics/mAP50-95(B)" in metrics.results_dict
+
+
+def test_val_save_csv(tmp_path) -> None:
+    """Test that save_csv=True writes per_class_metrics.csv after validation."""
+    import shutil
+
```
[the remark this answers is not in the document]
Comment (woreom) [reply tests/test_python.py]: I have read the CLA Document and I sign the CLA

**Comment by woreom** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
Thanks for this! I've pushed a rebase onto `main` and a fix for a header-alignment bug: with `iou_metrics` set, `SegmentMetrics`/`PoseMetrics` append the extra box-mAP columns (via `DetMetrics.keys`/`mean_results`) but `get_desc` was hardcoded to 10 columns, so the console header was shifted out of alignment. Fixed by inserting the extra columns between the Box and Mask/Pose blocks.

**Simplification suggestion (Core Principles)** 🧹 — in `metrics.py`, store grid **indices** instead of snapped float thresholds to remove the float→index→float→index round-trip. That lets you delete the `iou_metrics` `@property` getter + setter (~15 lines) and `_iou_idx()` (2 lines):

```python
self.iou_indices = sorted({min(9, max(0, round((min(0.95, max(0.5, t)) - 0.5) / 0.05))) for t in (value or [])})
```
then `keys`/`_extra_maps`/`class_result`/`get_desc` index `all_ap[:, i]` directly. Same outputs, ~20 fewer lines. (Minor: `save_csv` just wraps the already-public `metrics.to_csv()` — fine to keep, but it's a convenience flag, not new capability.)

**Comment by Jordan-Pierce** [comment]
Hi, realized looking at the training output that these metrics do not exist. Came looking and found this PR. Is this still on the roadmap?

**Comment by glenn-jocher** [comment]
This remains under consideration—the PR is open, but we can’t commit to a roadmap or merge timeline. Currently `mAP75` is available programmatically as `metrics.box.map75`, while configurable extra IoU metrics in training/validation output are not yet supported.

**Review by glenn-jocher — APPROVED** [review]
LGTM. Cold full-diff review of 967aecdc2505b81338b743163df54565586e42a7: exactly one added AP75 field at the existing summary owner. The IoU=0.75 column stays aligned with each class, and inherited summaries and all existing exports receive it without changing metric tuples or training-log keys. Real detection, segmentation, pose, and OBB validation verified the values in summary/DataFrame/CSV/JSON; empty summaries and Ruff checks passed. All previous findings concerned code removed by the simplification.

**Comment by glenn-jocher** [comment]
Thanks @woreom for the contribution and for highlighting the missing AP75 reporting. We simplified this PR to one line in the existing per-class summary, so AP75 now flows through summary, DataFrame, CSV, and JSON exports without new configuration arguments. Real detection, segmentation, pose, and OBB validation passed, and the final diff is LGTM. Merged at the maintainer's request while the ARM test job was still running; all other applicable checks had passed. Thank you!
